src/PostgREST/App.hs

```diff
@@ -404,15 +410,17 @@ parseRequest schema allRels rootTableName httpRequest reqBody =
               then M.fromList <$> (zip <$> flds <*> (head <$> vals))
               else Left "Expecting a sigle CSV line with header or a JSON object"
 
-createReadStatement :: SqlQuery -> Maybe NonnegRange -> Bool -> Bool -> B.Stmt P.Postgres
-createReadStatement selectQuery range countTable asCsv =
+createReadStatement :: SqlQuery -> Maybe NonnegRange -> Bool -> Bool -> Bool -> B.Stmt P.Postgres
+createReadStatement selectQuery range isSingle countTable asCsv =
   B.Stmt (
     wrapQuery selectQuery [
       if countTable then countAllF else countNoneF,
       countF,
       "null", -- location header can not be calucalted
-      if asCsv then asCsvF else asJsonF
-    ] selectStarF range
+      if asCsv
+        then asCsvF
+        else if isSingle then asJsonSingleF else asJsonF
+    ] selectStarF (if isNothing range && isSingle then Just $ singletonRange 0 else range)
```

Comment:
Just remembered, override is necessary because if the range specifies limit > 1 then asJsonSingleF will fail